libqpdf/QPDFValue.cc

```diff
@@ -43,3 +44,15 @@ QPDFValue::objectWarning(std::string const& warning) const
 {
     warn(QPDFExc(qpdf_e_object, "", getDescription(), 0, warning));
 }
+
+// Array methods
+
+int
+QPDFValue::size(bool type_warn, bool allow_null) const
+{
+    if (type_warn) {
```

Comment:
This makes me a little uncomfortable. Doing type checks inside QPDFValue makes me feel like we're slipping a little bit into QPDFObjectHandle land. If QPDFObject/QPDFValue is at least partly intended to be the basis for a more typesafe object handle, does it make more sense to just leave these checks in QPDFObjectHandle? That seems like the right layer to address type issues.

Reply:
To clarify, with this change, something like `asInteger()->size()` moves back from being a compile-time error to being a runtime error, and I thought one of the main thrusts here was to get rid of that kind of thing.

Reply:
It will always be a runtime error, asInteger will return a nullptr. Using virtual methods rather than dynamic casting is just a less expensive way of generating the runtime error .

Reply:
if asInteger() returns a pointer to a type that doesn't have a size() method, it will be a compile error, won't it? 
```c++
#include <memory>

class Base {
};

class D1: public Base {
  public:
    void f()
    {
    }
};

class D2: public Base {
};

void f() {
    auto b1 = std::make_shared<D1>();
    dynamic_cast<D1*>(b1.get())->f();
    dynamic_cast<D2*>(b1.get())->f();  // <-- compile error
}
```
To me, this is worth the overhead of dynamic cast. To code it correctly, you'd have to know the thing is an array or call isArray().

I suppose another way to do it that avoids dynamic cast is to use a child struct with one of each type.Each derived class's constructor could assign into this protected member variable in the base class. The code would be a little ugly for this, and each individual `as` would probably need to be done with template specialization, but it would avoid the dyanmic cast. It depends on what you're willing to trade to remove the dynamic cast overhead.

#include <memory>
#include <stdexcept>
#include <iostream>
#include <cassert>

class D1;
class D2;

class Base {
  public:
    template <typename T>
    T* as()
    {
        // Provide a template specialization for each derived class
        // that returns a value from child
        throw std::logic_error("no specialization provided");
    }

  protected:
    struct {
        // Each derived class must have a field here and set it to
        // `this` in its constructor.
        D1* d1{nullptr};
        D2* d2{nullptr};
    } child;
};

class D1: public Base {
  public:
    D1() {
        this->child.d1 = this;
    }

    void f()
    {
        std::cout << "D1::f" << std::endl;
    }
};

template <>
D1* Base::as()
{
    return child.d1;
}

class D2: public Base {
};

template <>
D2* Base::as()
{
    return child.d2;
}

int main() {
    auto b1 = std::make_shared<D1>();
    b1->as<D1>()->f();
//    b1->as<D2>()->f(); // <- compile error
    assert(b1->as<D2>() == nullptr);
    return 0;
}
```
It's kind of a C-style roll-your-own inheritance.

So I guess the questions, which of these would we rather see?
1. Overhead of RTTI for dynamic cast
2. Virtual methods (loss of compile-time checking for type-specific methods, distribution of error checking)
3. Extra boiler-plate in each class to keep type-safety and avoid RTTI

I guess I'd go with 1 or 3 depending on how much overhead RTTI has. Looking at my sample implementation, I don't think 3 is really all that bad from a code ugliness angle.

Reply:
(By `asInteger()` and `asArray()`, I mean `as<QPDF_Integer>()` and `as<QPDF_array>()`. I momentarily forgot they had been replaced by a templated function.)

Reply:
With this implementation, templated `as` isn't really natural. Here's a simpler version that reverts to the specific `as` methods:
```c++
#include <memory>
#include <stdexcept>
#include <iostream>
#include <cassert>

class D1;
class D2;

class Base {
  public:
    // Implement for each derived class to return the right value from child
    D1* asD1() {
        return child.d1;
    }
    D2* asD2() {
        return child.d2;
    }

  protected:
    struct {
        // Each derived class must have a field here and set it to
        // `this` in its constructor.
        D1* d1{nullptr};
        D2* d2{nullptr};
    } child;
};

class D1: public Base {
  public:
    D1() {
        this->child.d1 = this;
    }

    void f()
    {
        std::cout << "D1::f" << std::endl;
    }
};

class D2: public Base {
};

int main() {
    auto b1 = std::make_shared<D1>();
    b1->asD1()->f();
//    b1->as<D2>()->f(); // <- compile error
    assert(b1->asD2() == nullptr);
    return 0;
}
```